In [2]:
import pandas as pd
import numpy as np
import mlxtend
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.frequent_patterns import association_rules,apriori
import warnings
warnings.filterwarnings("ignore")    

## Read Data

In [5]:
df=pd.read_csv('https://github.com/thanrada301/Data/raw/main/basket_analysis.csv')
df.head()

Unnamed: 0  Apple  Bread  Butter  Cheese   Corn   Dill   Eggs  Ice cream  \
0           0  False   True   False   False   True   True  False       True   
1           1  False  False   False   False  False  False  False      False   
2           2   True  False    True   False  False   True  False       True   
3           3  False  False    True    True  False   True  False      False   
4           4   True   True   False   False  False  False  False      False   

   Kidney Beans   Milk  Nutmeg  Onion  Sugar  Unicorn  Yogurt  chocolate  
0         False  False   False  False   True    False    True       True  
1         False   True   False  False  False    False   False      False  
2         False   True   False  False  False    False    True       True  
3         False   True    True   True  False    False   False      False  
4         False  False   False  False  False    False   False      False

## Clean Data

In [6]:
df.drop('Unnamed: 0',axis=1,inplace=True)

In [7]:
df

Apple  Bread  Butter  Cheese   Corn   Dill   Eggs  Ice cream  \
0    False   True   False   False   True   True  False       True   
1    False  False   False   False  False  False  False      False   
2     True  False    True   False  False   True  False       True   
3    False  False    True    True  False   True  False      False   
4     True   True   False   False  False  False  False      False   
..     ...    ...     ...     ...    ...    ...    ...        ...   
994  False   True   False   False  False  False   True      False   
995   True  False   False   False   True  False  False      False   
996   True  False   False   False   True   True  False      False   
997  False  False    True    True   True  False   True       True   
998  False  False   False   False  False  False  False      False   

     Kidney Beans   Milk  Nutmeg  Onion  Sugar  Unicorn  Yogurt  chocolate  
0           False  False   False  False   True    False    True       True  
1           False   True   False  False  False    False   False      False  
2           False   True   False  False  False    False    True       True  
3           False   True    True   True  False    False   False      False  
4           False  False   False  False  False    False   False      False  
..            ...    ...     ...    ...    ...      ...     ...        ...  
994         False  False   False  False  False     True   False       True  
995          True   True    True  False  False    False    True      False  
996         False  False   False  False   True    False   False       True  
997          True  False    True  False   True    False    True       True  
998         False   True   False  False  False    False   False       True  

[999 rows x 16 columns]

In [8]:
df.shape

(999, 16)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Apple         999 non-null    bool 
 1   Bread         999 non-null    bool 
 2   Butter        999 non-null    bool 
 3   Cheese        999 non-null    bool 
 4   Corn          999 non-null    bool 
 5   Dill          999 non-null    bool 
 6   Eggs          999 non-null    bool 
 7   Ice cream     999 non-null    bool 
 8   Kidney Beans  999 non-null    bool 
 9   Milk          999 non-null    bool 
 10  Nutmeg        999 non-null    bool 
 11  Onion         999 non-null    bool 
 12  Sugar         999 non-null    bool 
 13  Unicorn       999 non-null    bool 
 14  Yogurt        999 non-null    bool 
 15  chocolate     999 non-null    bool 
dtypes: bool(16)
memory usage: 15.7 KB


In [11]:
df.isnull().sum()

Apple           0
Bread           0
Butter          0
Cheese          0
Corn            0
Dill            0
Eggs            0
Ice cream       0
Kidney Beans    0
Milk            0
Nutmeg          0
Onion           0
Sugar           0
Unicorn         0
Yogurt          0
chocolate       0
dtype: int64

In [12]:
df.mean()

Apple           0.383383
Bread           0.384384
Butter          0.420420
Cheese          0.404404
Corn            0.407407
Dill            0.398398
Eggs            0.384384
Ice cream       0.410410
Kidney Beans    0.408408
Milk            0.405405
Nutmeg          0.401401
Onion           0.403403
Sugar           0.409409
Unicorn         0.389389
Yogurt          0.420420
chocolate       0.421421
dtype: float64

## Model

In [14]:
frequent_itemsets = apriori(df,min_support =0.05,use_colnames=True)
frequent_itemsets.sort_values('support',ascending=False)

support                                  itemsets
15   0.421421                               (chocolate)
2    0.420420                                  (Butter)
14   0.420420                                  (Yogurt)
7    0.410410                               (Ice cream)
12   0.409409                                   (Sugar)
..        ...                                       ...
907  0.050050  (chocolate, Unicorn, Kidney Beans, Dill)
908  0.050050                (Sugar, Dill, Milk, Onion)
909  0.050050              (Unicorn, Dill, Milk, Onion)
763  0.050050         (Corn, Ice cream, Butter, Cheese)
848  0.050050            (Unicorn, Dill, Cheese, Onion)

[947 rows x 2 columns]

### association rules

In [21]:
# metrices = ['support', 'confidence', 'lift', 'leverage', and 'conviction']

metric='lift'
# rules = association_rules(frequent_itemsets, metric = "support", min_threshold = .2)
# rules = association_rules(frequent_itemsets, metric = "confidence", min_threshold = .5)
rules = association_rules(frequent_itemsets, metric = metric, min_threshold = 1)

rules.sort_values('lift', ascending = False, inplace = True)
rules

antecedents            consequents  antecedent support  \
5637         (Dill, Cheese)  (Kidney Beans, Onion)            0.177177   
5636  (Kidney Beans, Onion)         (Dill, Cheese)            0.170170   
6644     (chocolate, Onion)        (Unicorn, Dill)            0.196196   
6645        (Unicorn, Dill)     (chocolate, Onion)            0.168168   
5665   (Kidney Beans, Dill)    (chocolate, Cheese)            0.172172   
...                     ...                    ...                 ...   
458         (Yogurt, Apple)               (Cheese)            0.187187   
880         (Bread, Butter)                 (Dill)            0.180180   
881                  (Dill)        (Bread, Butter)            0.398398   
710   (Apple, Kidney Beans)                (Onion)            0.176176   
715                 (Onion)  (Apple, Kidney Beans)            0.403403   

      consequent support   support  confidence      lift  leverage  \
5637            0.170170  0.055055    0.310734  1.826022  0.024905   
5636            0.177177  0.055055    0.323529  1.826022  0.024905   
6644            0.168168  0.060060    0.306122  1.820335  0.027066   
6645            0.196196  0.060060    0.357143  1.820335  0.027066   
5665            0.186186  0.057057    0.331395  1.779914  0.025001   
...                  ...       ...         ...       ...       ...   
458             0.404404  0.076076    0.406417  1.004977  0.000377   
880             0.398398  0.072072    0.400000  1.004020  0.000289   
881             0.180180  0.072072    0.180905  1.004020  0.000289   
710             0.403403  0.071071    0.403409  1.000014  0.000001   
715             0.176176  0.071071    0.176179  1.000014  0.000001   

      conviction  zhangs_metric  
5637    1.203933       0.549768  
5636    1.216347       0.545126  
6644    1.198816       0.560648  
6645    1.250361       0.541757  
5665    1.217182       0.529307  
...          ...            ...  
458     1.003391       0.006093  
880     1.002669       0.004884  
881     1.000884       0.006656  
710     1.000010       0.000017  
715     1.000003       0.000024  

[7086 rows x 10 columns]

การที่เรียงข้อมูลตามค่า lift มาก แสดงว่ากลุ่มสินค้า antecedent นั้นมีความเกี่ยวข้องหรือมีความสัมพันธ์กับกลุ่มสินค้า consequent สูง  
เมื่อลูกค้าหยิบกลุ่มสินค้า antecedent ลงตะกร้าไปแล้ว มีโอกาสที่ลูกค้าจะหยิบกลุ่มสินค้า consequent เพิ่มมากขึ้นตามค่า lift นั้นๆ  
* ตัวอย่างเช่น  
    - เมื่อลูกค้าหยิบ Dill,Cheese โอกาสที่จะหยิบ Kidney Beans,Onion มีค่า 1.826002 เท่า  
    - เมื่อลูกค้าหยิบ Dill โอกาสที่จะหยิบ Bread,Butter มีค่า 1.004020 เท่า  

ดังนั้นหากต้องการให้ยอดขายเพิ่มขึ้นอีก ควรที่จะจัดสินค้าที่มีความสัมพันธ์สูงเอาไว้ด้วยกัน เพื่อที่จะให้ลูกค้ามองเห็นได้ง่ายและมีโอกาสที่ลูกค้าจะหยิบสินค้าไปด้วยกันมากขึ้น  หรือจะเป็นการจัดโปรโมชั่นระหว่างสินค้าที่มีความสัมพันธ์กันสูง เพื่อที่ง่ายต่อการตัดสินใจของลูกค้า